lib/CL/pocl_llvm_api.cc

```diff
@@ -1459,61 +1759,40 @@ static bool moduleReferencesImages(llvm::Module *M) {
 static PassManager &
 kernel_compiler_passes(cl_device_id device, llvm::Module *input,
                        const std::string &module_data_layout) {
-  static std::map<cl_device_id, PassManager*> kernel_compiler_passes;
 
-  bool SPMDDevice = device->spmd;
+  PassManager *Passes = nullptr;
+  PassRegistry *Registry = nullptr;
 
-  if (kernel_compiler_passes.find(device) != 
-      kernel_compiler_passes.end())
-    {
-      return *kernel_compiler_passes[device];
+  bool refs_images = moduleReferencesImages(input);
+
+  std::map<cl_device_id, PassManager *> &kernel_pass_map =
+      (refs_images ? kernel_passes_noflatten : kernel_passes_flatten);
```

Comment:
Can you try the no-flatten by default now? I think LLVM should take care of "sensible" inlining in standard opts.